source/core/ObjectRef.cpp

```diff
@@ -0,0 +1,142 @@
+}
+
+// Clean-up Proc, run when top level VI finishes, disposing object reference.
+static void CleanUpObjectReference(intptr_t arg) {
+    ObjectRefNum refnum = ObjectRefNum(arg);
+    ObjectReferenceDestroy(refnum);
+}
+
+// ObjectReferenceCreate -- create a object ref linked to object associated with objectTag on given VI.
+ObjectRefNum ObjectReferenceCreate(VirtualInstrument *vi, const StringRef &objectTag) {
+    ObjectRefInfo ObjectRefInfo(vi, objectTag);
+    ObjectRefNum refnum = ObjectRefNumManager::RefNumStorage().NewRefNum(&ObjectRefInfo);
+    ObjectRefNumManager::AddCleanupProc(null, CleanUpObjectReference, refnum);
+    return refnum;
+}
+
+// ObjectReferenceCreate -- RefNumVal version
+ObjectRefNum ObjectReferenceCreate(RefNumVal *pRefNumVal, VirtualInstrument *vi, const SubString &objectTag) {
+    STACK_VAR(String, objectTagStrVar);
+    StringRef objectTagStr = objectTagStrVar.DetachValue();
+    SubString ss(objectTag);
+    objectTagStr->AppendSubString(&ss);
+    ObjectRefNum refnum = ObjectReferenceCreate(vi, objectTagStr);
+    pRefNumVal->SetRefNum(refnum);
+    return refnum;
+}
+
+NIError ObjectReferenceLookup(ObjectRefNum refnum, VirtualInstrument **pVI, StringRef *pobjectTag) {
+    ObjectRefInfo ObjectRefInfo;
+    NIError err = ObjectRefNumManager::RefNumStorage().GetRefNumData(refnum, &ObjectRefInfo);
+    if (err == kNIError_Success) {
+        *pVI = ObjectRefInfo.vi;
+        *pobjectTag = ObjectRefInfo.objectTag;
+    } else {
+        *pVI = NULL;
+        *pobjectTag = NULL;
+    }
+    return err;
+}
+
+// ObjectReferenceAppendDescription -- output the VI and object a refnum is linked to (for test validation)
+NIError ObjectReferenceAppendDescription(StringRef str, ObjectRefNum refnum) {
+    ObjectRefInfo ObjectRefInfo;
+    NIError err = ObjectRefNumManager::RefNumStorage().GetRefNumData(refnum, &ObjectRefInfo);
+    if (err == kNIError_Success) {
+        SubString viName;
+        if (ObjectRefInfo.vi)
+            viName = ObjectRefInfo.vi->VIName();
+        str->Append('<');
+        str->AppendSubString(&viName);
+        str->Append(',');
+        str->AppendStringRef(ObjectRefInfo.objectTag);
+        str->Append('>');
+    }
+    return err;
+}
+
+VIREO_FUNCTION_SIGNATURE2(IsNotAObjectRefnum, RefNumVal, Boolean)
+{
+    RefNumVal* refnumPtr = _ParamPointer(0);
+    if (!refnumPtr || ObjectRefNumManager::RefNumStorage().GetRefNumData(refnumPtr->GetRefNum(), NULL) != kNIError_Success)
+        _Param(1) = true;
+    else
+        _Param(1) = false;
+    return _NextInstruction();
+}
+
+VIREO_FUNCTION_SIGNATURE3(IsEQRefnum, RefNumVal, RefNumVal, Boolean);
+VIREO_FUNCTION_SIGNATURE3(IsNERefnum, RefNumVal, RefNumVal, Boolean);
+
+//------------------------------------------------------------
+DEFINE_VIREO_BEGIN(ObjectRefs)
+    DEFINE_VIREO_REQUIRE(VirtualInstrument)
+
+    DEFINE_VIREO_TYPE(ObjectRefNumInfo, "c(e(VirtualInstrument vi) e(String objectTag))")
```

Comment:
Don't think the ObjectRef should have a reference to the VI...

Reply:
Why not? The implementation parallels control ref, which makes sense to me.